### PR TITLE
fix: solves requests and responses types and adds typesafety 🧙‍

### PR DESCRIPTION
**Pull request by miguelvelasquezdev**
- Fix type error in packages/uploadthing/src/internal/handler.ts

Screenshot 2023-05-27 at 7 25 35 PM

- add typesafety to requests and possible responses

---

- [x] Funcitonal tests

### CODE REVIEW

**Comment by juliusmarminge** [comment]
Hey, thanks for this. We just merged a different PR for this thst simplifierad the types a bit. Can you merge that in and resolve the conflicts?

**Comment by juliusmarminge** [comment]
Tests failing

File: packages/uploadthing/src/types.ts [hunk 1]
```diff
@@ -52,29 +52,29 @@ type ResolverOptions<TParams extends AnyParams> = {
...
+      req: RequestAdapter<TRuntime>,
+      res: ResponseAdapter<TRuntime>
     ) => MaybePromise<TOutput>;
 
-export type ReqMiddlewareFn<TOutput extends Record<string, unknown>> =
-  MiddlewareFn<TOutput, "web">;
-export type NextReqMiddlewareFn<TOutput extends Record<string, unknown>> =
-  MiddlewareFn<TOutput, "app">;
-export type NextApiMiddlewareFn<TOutput extends Record<string, unknown>> =
-  MiddlewareFn<TOutput, "pages">;
```
Comment (juliusmarminge) [inline packages/uploadthing/src/types.ts:76]: not sure how many are using these but it's technically a breaking change removing these exports

File: packages/uploadthing/src/types.ts [hunk 2]
```diff
@@ -52,29 +52,29 @@ type ResolverOptions<TParams extends AnyParams> = {
...
-  ? (req: MiddlewareFnArgs<TRuntime>["req"]) => MaybePromise<TOutput>
+  TRuntime extends string
+> = ResponseAdapter<TRuntime> extends undefined
+  ? (req: RequestAdapter<TRuntime>) => MaybePromise<TOutput>
   : (
-      req: MiddlewareFnArgs<TRuntime>["req"],
-      res: MiddlewareFnArgs<TRuntime>["res"],
+      req: RequestAdapter<TRuntime>,
+      res: ResponseAdapter<TRuntime>
     ) => MaybePromise<TOutput>;
```
Comment (juliusmarminge) [inline packages/uploadthing/src/types.ts:76]: im not convinced this is better

File: packages/uploadthing/src/internal/handler.ts [hunk 3]
```diff
@@ -230,7 +230,7 @@ export const buildRequestHandler = <
 
     const reqBody = (await req.json()) as {
       file: UploadedFile;
-      files: unknown;
+      files: string[];
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:233]: unknown is more accurate

File: packages/uploadthing/src/internal/handler.ts [hunk 4]
```diff
@@ -253,8 +253,7 @@ export const buildRequestHandler = <
     }
 
     try {
-      const { files } = reqBody as { files: string[] };
-      // @ts-expect-error TODO: Fix this
+      const { files } = reqBody;
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:256]: files should still be unknown here really... we validate it further down (L260).

I'm guessing the previous `as` here is to support the destructuring but maybe just skip that?

```ts
try {
  // ...
  if (!Array.isArray(reqBody.files) || !reqBody.files.every((f) => typeof f === "string"))
        throw new Error("Need file array");
  // ✅ string[]
```

File: packages/uploadthing/src/types.ts (hunk 1 above)
Comment (miguelvelasquezdev) [reply packages/uploadthing/src/types.ts:76]: it is exported, but it's not used anywhere...

File: packages/uploadthing/src/internal/handler.ts (hunk 4 above)
Comment (miguelvelasquezdev) [reply packages/uploadthing/src/internal/handler.ts:256]: so the `as` is unnecessary?

File: packages/uploadthing/src/types.ts (hunk 1 above)
Comment (miguelvelasquezdev) [reply packages/uploadthing/src/types.ts:76]: Ohh, i got it

File: packages/uploadthing/src/types.ts (hunk 2 above)
Comment (miguelvelasquezdev) [reply packages/uploadthing/src/types.ts]: I made changes in the last commit, maybe it's better to read

File: packages/uploadthing/src/upload-builder.test.ts [hunk 5]
```diff
@@ -62,7 +62,7 @@ it("passes `Request` by default", () => {
   const f = createBuilder();
 
   f(["image"]).middleware((req) => {
-    expectTypeOf(req).toMatchTypeOf<Request>();
+    expectTypeOf(req).toMatchTypeOf<Partial<Request> & RequestBase>();
```
Comment (juliusmarminge) [inline packages/uploadthing/src/upload-builder.test.ts:65]: I don't like if this is exposed to the user - they should just see a `Request` / `NextRequest` or `NextApiRequest` etc

**Pull request closed without merging**